services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/Persistence.kt

```diff
@@ -197,3 +199,31 @@ interface HankeRepository : JpaRepository<HankeEntity, Int> {
 
 }
 
+enum class CounterType {
+    HANKETUNNUS
+}
+
+@Entity
+@Table(name = "idcounter")
+class IdCounter(
+    @Id
+    @Enumerated(EnumType.STRING)
+    var counterType: CounterType? = null,
+    var value: Long? = null
+)
+
+interface IdCounterRepository : JpaRepository<IdCounter, CounterType> {
+    /*
+    This SQL clause has some PostgreSQL specific thingies:
+    'FOR UPDATE' in nested SELECT clause makes sure that no other process can update the row during this whole UPDATE clause
+    'RETURNING' in the end is for UPDATE clase to return not just the number of affected rows but also the column data of those rows (a single row in our case)
+    With these two specialities we can assure that concurrent calls for this method will never return duplicate values.
+    Notice also that the method returns a list even though there is always only max. 1 item in it
+     */
+    @Modifying(clearAutomatically = true, flushAutomatically = true)
+    @Query(
+        "UPDATE IdCounter SET value = (SELECT value FROM IdCounter WHERE counterType = :counterType FOR UPDATE) + 1 WHERE counterType = :counterType RETURNING *",
+        nativeQuery = true
+    )
+    fun incrementAndGet(counterType: String): List<IdCounter>
+}
```

Comment:
Found my missing piece! This seems to do the trick of preventing two simultaneuous calls.

Reply:
Yes.. but adding the annual nullfication might make things trickier..